### PR TITLE
Added AutoCloseable tests for EntityManagerFactory and EntityManager.

### PR DESCRIPTION
**Pull request by Tomas-Kraus**
Signed-off-by: Tomas Kraus <tomas.kraus@oracle.com>

---
name: Pull Request
about: Create a pull request for a Platform TCK change
title: 'Added AutoCloseable tests for EntityManagerFactory and EntityManager'
labels: ''
assignees: ''

---

**Fixes Issue**
https://github.com/eclipse-ee4j/jakartaee-tck/issues/730

**Related Issue(s)**
N/A

**Describe the change**
Added autoCloseableTest into EMF anf EM test clients.

**Additional context**
N/A

CC @alwin-joseph @anajosep @arjantijms @cesarhernandezgt @dblevins @m0mus @edbratt @gurunrao @jansupol @jgallimore @kazumura @kwsutter @LanceAndersen @bhatpmk @RohitKumarJain @shighbar @gthoman @brideck @scottmarlow

### CODE REVIEW

File: src/com/sun/ts/tests/jpa/core/entityManager/Client.java [hunk 1]
```diff
@@ -276,6 +277,43 @@ public boolean verifyListEmployees(List<Employee> expected,
...
+      } finally {
+        if (em.isOpen() == true) {
+          TestUtil.logErr("EntityManager isOpen() returned true outside try block");
+          pass = false;
+        }
+      }
+    } catch (Throwable t) {
+      throw new Fault("autoCloseableTest failed with Exception", t);
+    }
+    if (!pass) {
```
[the remark this answers is not in the document]
Comment (LanceAndersen) [reply src/com/sun/ts/tests/jpa/core/entityManager/Client.java:312]: I agree that this test could be cleaned up with a minor rework to make it clearer and should be considered when updates are made to specific tests

[the remark this answers is not in the document]
Comment (LanceAndersen) [reply src/com/sun/ts/tests/jpa/core/entityManager/Client.java:312]: I think it is time to clean these up as they are updated.  use something like assertTrue/False/NotNull. etc making the code more readable

File: src/com/sun/ts/tests/jpa/core/entityManagerFactory/Client.java [hunk 2]
```diff
@@ -96,12 +98,49 @@ public void cleanup() throws Fault {
...
+    boolean pass = true;
+    EntityManagerFactory emf = null;
+    try (final EntityManagerFactory emfLocal
+                 = Persistence.createEntityManagerFactory(getPersistenceUnitName())) {
+      emf = emfLocal;
+      if (emf == null) {
+        TestUtil.logErr("createEntityManagerFactory(String) returned a null result");
+        pass = false;
+      }
+      if (emf.isOpen() == false) {
```
[the remark this answers is not in the document]
Comment (Tomas-Kraus) [reply src/com/sun/ts/tests/jpa/core/entityManagerFactory/Client.java]: Changed the test to throw Fail exception right away.

**Comment by Tomas-Kraus** [comment]
I'll keep this test as draft until asserttion IDs are resolved properly. Asked Alwin for some help with this.

**Comment by scottmarlow** [comment]
> I'll keep this test as draft until assertion IDs are resolved properly. Asked Alwin for some help with this.

For reference https://github.com/eclipse-ee4j/jakartaee-tck/wiki/Steps-to-generate-the-html-version-of-the-SpecAssertions.xml describes how to generate the assertions.html version of a SpecAssertions.xml input file but I'm not sure of when we need to run that versus manually updating the assertions.html files.  

@lukasj should these new AutoCloseable tests use a JavaDoc (for https://github.com/eclipse-ee4j/jpa-api/blob/master/api/src/main/java/jakarta/persistence/EntityManager.java + https://github.com/eclipse-ee4j/jpa-api/blob/master/api/src/main/java/jakarta/persistence/EntityManagerFactory.java) or SPEC (for https://github.com/eclipse-ee4j/jpa-api/blob/master/spec/src/main/asciidoc/ch07-entitymanagers-and-persistence-contexts.adoc) assertion ID?

**Comment by alwin-joseph** [comment]
The assertions in the test description seems to be **api assertions** or javadoc ids. I have added a script in https://github.com/eclipse-ee4j/jakartaee-tck-tools/pull/9/ that was used to generate the javadoc assertions from javadocs (Thanks to Jan). Tomas would help trying this script to generate the javadoc ids for jpa. 

IMO it would be sufficient to use the api assertion for the existing apis that were changed in this PR. Finding the right javadoc assertion id for the changed methods is required. https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/install/jpa/docs/assertions/api/api-assertions-3_0.html has the api assertions for the previous release. 

> For reference https://github.com/eclipse-ee4j/jakartaee-tck/wiki/Steps-to-generate-the-html-version-of-the-SpecAssertions.xml describes how to generate the assertions.html version of a SpecAssertions.xml input file but I'm not sure of when we need to run that versus manually updating the assertions.html files.

The https://github.com/eclipse-ee4j/jakartaee-tck/wiki/Steps-to-generate-the-html-version-of-the-SpecAssertions.xml only provides instructions to generate the html version of the already existing spec assertion xml files.

**Comment by scottmarlow** [comment]
> @scottmarlow Do you have a pointer to docs describing what these assertions are good for/how they should be written?

I'll draft something up on https://github.com/eclipse-ee4j/jakartaee-tck/wiki/Guide-to-adding-Specification-Assertions and start a discussion thread for the same on the Platform TCK ML.

**Comment by Tomas-Kraus** [comment]
I tried to create new API assertions list for 3.1.0 but there was no information about `EntityManager/EntityManagerFactory extends AutoCloseable`. I have no idea how to add such ID into the list.
Maybe I can just reuse old IDs for EntityManager/EntityManagerFactory `close()` method which is there since the beginning.

**Comment by Tomas-Kraus** [comment]
Just keep in mind, that this is JPA 3.1 modification and [JPA 3.1](https://github.com/eclipse-ee4j/jpa-api/releases/tag/3.1-3.1.0-RC1-RELEASE) is needed for build/run. Artifacts are available at https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/persistence/jakarta.persistence-api/3.1.0-RC1/

**Comment by scottmarlow** [comment]
> I tried to create new API assertions list for 3.1.0 but there was no information about `EntityManager/EntityManagerFactory extends AutoCloseable`. I have no idea how to add such ID into the list. Maybe I can just reuse old IDs for EntityManager/EntityManagerFactory `close()` method which is there since the beginning.

I built the JPA 3.1 spec source locally and see `AutoCloseable` mentioned in a few files.  In no particular order, below are some SPEC references:

From jpa-api/spec/target/generated-docs/jakarta-persistence-spec-3.1-SNAPSHOT.html#jakarta-persistence-3-1:

```
A.3. Jakarta Persistence 3.1

EntityManagerFactory and EntityManager interfaces extend java.lang.AutoCloseable interface
```

From jpa-api/spec/target/generated-docs/jakarta-persistence-spec-3.1-SNAPSHOT.html#a1066 (`3.1.1. EntityManager Interface`)
```
public interface EntityManager extends AutoCloseable {
```

From file:///home/smarlow/work/jakarta/jpa-api/spec/target/generated-docs/jakarta-persistence-spec-3.1-SNAPSHOT.html#entitymanagerfactory-interface (`7.4. EntityManagerFactory Interface`):
```
public interface EntityManagerFactory extends AutoCloseable {
```

I also looked at javadoc via https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/persistence/jakarta.persistence-api/3.1.0-RC1/jakarta.persistence-api-3.1.0-RC1-javadoc.jar to see possible options for assertions based on that:

The jakarta.persistence/jakarta/persistence/EntityManagerFactory.html has:
```
All Superinterfaces:
    AutoCloseable

public interface EntityManagerFactory
extends AutoCloseable
```
The jakarta.persistence/jakarta/persistence/EntityManager.html has:
```
All Superinterfaces:
    AutoCloseable

public interface EntityManager
extends AutoCloseable
```

I suggest that api-assertions-3_0.html should have two new rows at the end added with:
```
ID=PERSISTENCE:JAVADOC:3491

Return=

Method/Field=

Description=EntityManagerFactory must extend java.lang.AutoCloseable

Required=true

Deprecated=

Testable=true
```

And:
```
ID=PERSISTENCE:JAVADOC:3492

Return=

Method/Field=

Description=EntityManager must extend java.lang.AutoCloseable

Required=true

Deprecated=

Testable=true
```

The internal/docs/jpa/Persistence31JavadocAssertions.xml should be updated to include the new IDs (I think the tests are reusing already allocated IDs that verify closing of EntityManager or EntityManagerFactory).  For the description field, it would be good to say more than I did, if anyone feels creative (perhaps based on wording in https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html).

File: src/com/sun/ts/tests/jpa/core/entityManager/Client.java [hunk 3]
```diff
@@ -276,6 +277,38 @@ public boolean verifyListEmployees(List<Employee> expected,
     return result;
   }
 
+  /*
+   * @testName: autoCloseableTest
+   *
+   * @assertion_ids: PERSISTENCE:JAVADOC:36
```
Comment (scottmarlow) [inline src/com/sun/ts/tests/jpa/core/entityManager/Client.java:283]: @Tomas-Kraus do you agree with my suggestion in https://github.com/eclipse-ee4j/jakartaee-tck/pull/759#issuecomment-966435181 that we should use a higher number than `36`? 

I think that we should add to the highest ID used currently, so perhaps something like `PERSISTENCE:JAVADOC:3491`.  

Thoughts?

**Comment by scottmarlow** [comment]
Looks like we should also increment `next-available-id` in https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/internal/docs/jpa/Persistence22JavadocAssertions.xml#L22 as part of this change.

File: src/com/sun/ts/tests/jpa/core/entityManager/Client.java (hunk 3 above)
Comment (scottmarlow) [reply src/com/sun/ts/tests/jpa/core/entityManager/Client.java:283]: Will use the current highest number `3491` and increment.  I'll make the change.

**Comment by scottmarlow** [comment]
I'm not sure why PERSISTENCE:JAVADOC:29 is now PERSISTENCE:JAVADOC:29__OLD in internal/docs/jpa/Persistence31JavaDocAssertions.html, @Tomas-Kraus was that a hint that some action needs to be done to the https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/src/com/sun/ts/tests/jpa/core/annotations/entity/Client.java#L65 source that references `PERSISTENCE:JAVADOC:29`?

**Comment by alwin-joseph** [comment]
> I'm not sure why PERSISTENCE:JAVADOC:29 is now PERSISTENCE:JAVADOC:29__OLD in internal/docs/jpa/Persistence31JavaDocAssertions.html, @Tomas-Kraus was that a hint that some action needs to be done to the https://github.com/eclipse-ee4j/jakartaee-tck/blob/master/src/com/sun/ts/tests/jpa/core/annotations/entity/Client.java#L65 source that references `PERSISTENCE:JAVADOC:29`?
> 
> It looks like you corrected some entries and kept the original entries with the `__OLD` name, which seems fine.

@Tomas-Kraus The entries with the `__OLD` name can be removed . We did the same for jaxrs , see https://github.com/eclipse-ee4j/jaxrs-api/pull/1059#issuecomment-989700203 

cc @jansupol

**Comment by scottmarlow** [comment]
> @Tomas-Kraus The entries with the `__OLD` name can be removed . We did the same for jaxrs , see [eclipse-ee4j/jaxrs-api#1059 (comment)](https://github.com/eclipse-ee4j/jaxrs-api/pull/1059#issuecomment-989700203)
> 
> cc @jansupol

I'll push a change to remove the __OLD entries.

**Comment by Tomas-Kraus** [comment]
PR #793 contains spec assertion IDs update for 3.1. I changed assertion IDs in AutoCloseable tests to point to 
- PERSISTENCE:SPEC:2517 - EntityManagerFactory and EntityManager interfaces extend java.lang.AutoCloseable interface

File: src/com/sun/ts/tests/jpa/core/entityManager/Client.java (hunk 3 above)
Comment (Tomas-Kraus) [reply src/com/sun/ts/tests/jpa/core/entityManager/Client.java]: I used spec reference instead. API lists are being generated directly from API build using script so manual changes are not a good idea here.

**Review by scottmarlow — APPROVED** [review]

**Pull request merged**